x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredModelSnapshotsRemover.java

```diff
@@ -208,65 +219,53 @@ public void onResponse(SearchResponse searchResponse) {
+                                (shots) -> snapshots.add(shots.result),
+                                (failure) ->
+                                    LOGGER.warn(new ParameterizedMessage("[{}] failed to find snapshot [{}]", jobId, snapshotId), failure)
+                                );
                         }
                     }
-                    deleteModelSnapshots(new VolatileCursorIterator<>(snapshotIds), listener);
+                    deleteModelSnapshots(snapshots, jobId, listener);
                 } catch (Exception e) {
                     onFailure(e);
                 }
             }
 
             @Override
             public void onFailure(Exception e) {
-                listener.onFailure(new ElasticsearchException("[" + jobId +  "] Search for expired snapshots failed", e));
+                listener.onFailure(new ElasticsearchException("[{}] Search for expired snapshots failed", e, jobId));
             }
         };
     }
 
-    private void deleteModelSnapshots(Iterator<JobSnapshotId> modelSnapshotIterator, ActionListener<Boolean> listener) {
-        if (modelSnapshotIterator.hasNext() == false) {
+    private void deleteModelSnapshots(List<ModelSnapshot> modelSnapshots, String jobId, ActionListener<Boolean> listener) {
+        if (modelSnapshots.isEmpty()) {
             listener.onResponse(true);
             return;
         }
-        JobSnapshotId idPair = modelSnapshotIterator.next();
-        DeleteModelSnapshotAction.Request deleteSnapshotRequest =
-            new DeleteModelSnapshotAction.Request(idPair.jobId, idPair.snapshotId);
-        deleteSnapshotRequest.setParentTask(getParentTaskId());
-        client.execute(DeleteModelSnapshotAction.INSTANCE, deleteSnapshotRequest, new ActionListener<>() {
-                @Override
-                public void onResponse(AcknowledgedResponse response) {
-                    try {
-                        deleteModelSnapshots(modelSnapshotIterator, listener);
-                    } catch (Exception e) {
-                        onFailure(e);
-                    }
-                }
+        JobDataDeleter deleter = new JobDataDeleter(client, jobId);
+        deleter.deleteModelSnapshots(modelSnapshots, ActionListener.wrap(
+            bulkResponse -> {
+                String msg = Messages.getMessage(
+                    Messages.JOB_AUDIT_SNAPSHOTS_DELETED,
+                    modelSnapshots.stream().map(ModelSnapshot::getSnapshotId),
+                    modelSnapshots.stream().map(ModelSnapshot::getDescription));
 
-                @Override
-                public void onFailure(Exception e) {
-                    listener.onFailure(new ElasticsearchException("[" + idPair.jobId +  "] Failed to delete snapshot ["
-                            + idPair.snapshotId + "]", e));
-                }
-            });
+                auditor.info(jobId, msg);
+                LOGGER.debug(() -> new ParameterizedMessage("[{}] {}", jobId, msg));
```

Comment:
Even though the audit message can be made less verbose, the debug should continue to list every ID and description, just in case something ever goes wrong with which snapshots are getting deleted.

Reply:
no doubt!